### PR TITLE
dialects: (acc) Add acc dialect and acc-cse pass to simplify setups

### PR DESCRIPTION
**Pull request by AntonLydike**
`acc` dialect: (snax) accelerator setup and dispatch code, used to represent setup, launching and waiting for accelerator calls.

`acc-cse` pass: elide redundant setup operations from IR.

### CODE REVIEW

File: compiler/dialects/acc.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+)
+
+
+@irdl_attr_definition
+class TokenType(ParametrizedAttribute, TypeAttribute):
+    """
+    Async token type for launched accelerator requests.
+    """
+
+    name = "acc.token"
```
Comment (jorendumoulin) [inline compiler/dialects/acc.py:33]: The `TokenType` is empty: let me just check if I get the working of this correctly: The sole purpose of a token is to easily know which `acc.wait` and `acc.launch` are linked together, which is useful in some passes? This won't actually be something that is lowered, as the `acc.await` will be based on the accelerator id, not the token.

File: compiler/dialects/acc.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    def __init__(
+        self,
+        vals: Sequence[SSAValue],
+        param_names: Sequence[str],
+        accelerator: str | StringAttr,
+        in_state: SSAValue | Operation | None = None,
+    ):
+        if not isinstance(accelerator, StringAttr):
+            accelerator = StringAttr(accelerator)
+
```
Comment (jorendumoulin) [inline compiler/dialects/acc.py:117]: Adding support for existing StringAttr params can be nice
```suggestion
        param_names: Sequence[str | StringAttr],
        accelerator: str | StringAttr,
        in_state: SSAValue | Operation | None = None,
    ):
        if not isinstance(accelerator, StringAttr):
            accelerator = StringAttr(accelerator)
        
        param_names = [
            StringAttr(param_name) if not isinstance(param_name, StringAttr) 
            else param_name 
            for param_name in param_names
        ]
```

File: compiler/transforms/acc_cse.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            (name, val) for name, val in op.iter_params() if prev_state.get(name) != val
+        ]
+
+        # Step 3: If no new params remain, yeet the whole op
+        if not new_params:
+            op.out_state.replace_by(op.in_state)
+            rewriter.erase_matched_op()
+            return
+
+        # Step 4: If all parameters change, we are a no-op
```
Comment (jorendumoulin) [inline compiler/transforms/acc_cse.py:53]:
```suggestion
        # Step 4: If all parameters change, do nothing
```

File: compiler/transforms/acc_cse.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    def match_and_rewrite(self, op: acc.SetupOp, rewriter: PatternRewriter, /):
+        # Step 1: Figure out previous state
+        prev_state = get_state_before(op)
+
+        # Step 2: Yeet out redundant stuff
+        new_params: list[tuple[str, SSAValue]] = [
+            (name, val) for name, val in op.iter_params() if prev_state.get(name) != val
+        ]
+
+        # Step 3: If no new params remain, yeet the whole op
```
Comment (jorendumoulin) [inline compiler/transforms/acc_cse.py:47]:
```suggestion
        # Step 3: If no parameters change, the whole op can be erased
```

File: compiler/transforms/acc_cse.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    return state
+
+
+class YeetSetupStuff(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: acc.SetupOp, rewriter: PatternRewriter, /):
+        # Step 1: Figure out previous state
+        prev_state = get_state_before(op)
+
+        # Step 2: Yeet out redundant stuff
```
Comment (jorendumoulin) [inline compiler/transforms/acc_cse.py:42]:
```suggestion
        # Step 2: Yeet out redundant stuff (= params that get set to the same value)
```

File: compiler/transforms/acc_cse.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+        for name, val in parent.iter_params():
+            if name not in state:
+                state[name] = val
+
+        op = parent
+    return state
+
+
+class YeetSetupStuff(RewritePattern):
```
Comment (jorendumoulin) [inline compiler/transforms/acc_cse.py:36]: I love the name, but perhaps `AccSetupCse`, is a tiny bit more clear?

**Review by jorendumoulin — APPROVED** [review]
Looks good! I have some minor suggestions

File: compiler/dialects/acc.py (hunk 1 above)
Comment (AntonLydike) [reply compiler/dialects/acc.py:33]: Yes, pretty much. The token is used so we can trace which acccelerator to wait on. E.g. the launch launches an accelerator, and this way the await will always match the launch accelerator.

This also allows us to maybe infer the kernel runtimes, so we can push setup calls up before the previous await, saving more cycles!

File: compiler/dialects/acc.py (hunk 2 above)
Comment (AntonLydike) [reply compiler/dialects/acc.py:117]: Yeah, we had this before, but pyright can't infer the types. We can add it, but it's not needed so far, and the way we ended up doing it in xDSL was:

```py
param_names : Sequence[str] | Sequence[StringAttr]  # cannot be a mixed sequence

# ...
params = tuple(param_names) # `isa` doesn't support Sequence, so cast to tuple
if not isa(params, tuple[StringAttr, ...]):
    params = tuple(StringAttr(x) for x in params
```
But this is not really nice looking

**Pull request merged**